sisyphus/job.py

```diff
@@ -651,9 +656,96 @@ def _sis_all_path_available(self):
+        Job._sis_get_expected_file_sizes.
+
+        """
+        stats = []
+        below_work = pathlib.Path(self._sis_path(gs.WORK_DIR)).rglob("*")
+        below_output = pathlib.Path(self._sis_path(gs.OUTPUT_DIR)).rglob("*")
+        for p in chain(below_work, below_output):
+            if p.is_file():
+                stat = p.stat()
+                rel_path = str(p.relative_to(self._sis_path()))
+                stats.append((rel_path, stat.st_mtime, stat.st_size))
+
+        return stats
+
+
+    @staticmethod
+    def _sis_get_expected_file_sizes(job: Union[str,"Job"], task: str = None,
+                                     timeout = gs.MAX_WAIT_FILE_SYNC) -> Dict[str, int]:
+        """
+        Tries to obtain the expected file sizes for files below `output` and `work` from the usage
+        files of the given job (or job dir). Returns None if the job had already been cleaned up.
+
+        If a usage file does not contain the file size information, this is either because the
+        respective task is still runnung or because the usage file is not yet synced. In this case,
+        retry until timeout and raise a TimeoutError.
+
+        When accumulating the information from several files, the most recent size info is retained
+        for every *existing* path. That is, deleted files are not part of the returned list.
+
+        The file paths returned are relative to the experiment directory.
+
+        If `task` is given, only usage files from these tasks are read.
+
+        """
+        # The job might be a Job object or the job directory
+        try:
+            job_dir = job._sis_path()
+        except AttributeError:
+            job_dir = job
+
+        if os.path.exists(os.path.join(job_dir, gs.JOB_FINISHED_ARCHIVE)):
+            logging.info("No expected file size info for job %s, is has already been cleaned up.", job_dir)
+            return None
+
+        m_times = defaultdict(int)
+        sizes = dict()
+        
+        exp = "{0}.{1}.*".format(gs.PLOGGING_FILE, task if task else "*")
+        for fn in pathlib.Path(job_dir).glob(exp):
+            start = time.time()
+            while True:
+                with open(fn) as f:
+                    try:
+                        stats = literal_eval(f.read())["file_stats"]
+                    except KeyError:
+                        # Fairly unlikely to happen: A job from an earlier sisyphus run should be cleaned up.
+                        logging.warning("%s contains no file_stats (was created by an older version of sisyphus).")
+                        stats = []
+                        break
+                
+                if stats:
+                    break
+                if time.time() - start > timeout:
+                    logging.error("%s not synced for more than %ds, file_stats still empty.", fn, timeout)
+                    raise TimeoutError
+                logging.info("%s not synced yet, file_stats still empty.", fn)
```

Comment:
This will cause problems if a path is available and should be used before the job is finished e.g. training of a neural model.
We could require that these paths have a special attribute set. They could then be excluded from this check.